R/glossary.R

```diff
@@ -0,0 +1,390 @@
+#'
+#' @param file path to the tex or yaml glossary file.
+#' @param format the glossary file format; could be either `"tex"` or `"yaml"`
+#' and will be inferred by default from the extension of `file`.
+#'
+#' @return
+#' A named list of glossary definitions with class `tex_glossary` and
+#' `glossary`. The list names are acronym entry labels (i.e., what you would
+#' pass to `\gls{}` when writing a tex document).
+#'
+#'
+#' @details
+#' For the tex format, the glossary file is expected to contain acronym
+#' entries with the form
+#'
+#' ```
+#' \newacronym{label}{abbreviation}{definition}
+#' ```
+#'
+#' or
+#'
+#' ```
+#' \newacronym[options]{label}{abbreviation}{definition}
+#' ```
+#'
+#' For the yaml format, the glossary file is expected to be able to be
+#' read with [yaml_as_df()]; the outer level should be the `label` and
+#' columns should be provided with the names `def` (definition) and
+#' `abb` (abbreviation). See **Examples**.
+#'
+#'
+#'
+#' @examples
+#' texfile <- system.file("glo", "glossary.tex", package = "pmtables")
+#'
+#' x <- read_glossary(texfile)
+#'
+#' head(x)
+#'
+#' x$WT
+#'
+#' x$SC
+#'
+#' yamfile <- system.file("glo", "glossary.yaml", package = "pmtables")
+#'
+#' y <- read_glossary(yamfile)
+#' head(y)
+#'
+#' # Format of the tex glossary file
+#' head(readLines(texfile))
+#'
+#' # Format of the yaml glossary file
+#' head(readLines(yamfile))
+#'
+#' yaml_as_df(yamfile)
+#'
+#'
+#' @seealso [as_glossary()]
+#' @export
+read_glossary <- function(file, format = guess_glo_fmt(file)) {
+  if(!file.exists(file)) {
+    abort(glue("Glossary file {file} does not exist."))
+  }
+  if(!format %in% c("tex", "yaml", "yml")) {
+    abort("format should be either tex, yaml, or yml")
+  }
+  if(format == "tex") {
+    ans <- read_tex_glossary(file)
+  } else {
+    ans <- read_yaml_glossary(file)
+  }
+  class(ans) <- c("tex_glossary", "glossary", "list")
+  ans
+}
+
+read_yaml_glossary <- function(file) {
+  parsed <- try(yaml_as_df(file, row_var = "label"), silent = TRUE)
+  if(inherits(parsed, "try-error")) {
+    abort(
+      c("Failed to parse glossary file; see ?yaml_as_df() for help formatting this file.",
+        as.character(parsed)
+      )
+    )
+  }
+  if(!all(c("label", "abb", "def") %in% names(parsed))) {
+    abort("yaml/yml glossary format requires names label, abb, and def.")
+  }
+  labels <- parsed[["label"]]
+  parsed <- select(parsed, "abb", "def")
+  data <- split(parsed, seq(nrow(parsed)))
+  data <- lapply(data, unlist)
+  data <- lapply(data, as_glossary_entry)
+  data <- lapply(data, setNames, c("abbreviation", "definition"))
+  names(data) <- labels
+  data
+}
+
+read_tex_glossary <- function(file) {
+  txt <- readLines(file, warn = FALSE)
+  if(!length(txt)) abort("The glossary file was empty.")
+  parse_tex_glossary(txt)
+}
+
+parse_tex_glossary <- function(txt) {
+  txt <- trimws(txt)
+  txt <- txt[grepl("^\\\\newacronym", txt)]
+  m <- regexec("\\{(.+)\\}\\{(.+)\\}\\{(.+)\\}.*$", txt)
```

Comment:
I suspect `newacronym` items permit a fair amount of flexibility in spaces and newlines (e.g., between the brackets for different arguments or at the start of the line). The parser here doesn't allow that, but I think that's okay given that the expected forms are clearly documented.

Beyond whitespace, it looks like extra brackets could trip up the parser. Here's an example of brackets in a trailing comment causing issues:

```R
parse_tex_glossary("\\newacronym{a}{b}{c} % foo}")
#> $a
#> c} % foo (b)
```

Perhaps this parser should strip the trailing comments from the line before doing the match.

Reply:
Thanks, Kyle. 

1. spaces at the start of a line are handled by `trimws()`
2. Modified the regex to allow for spaces between brackets
3. Trimmed parsed result to take care of spaces inside brackets at the start or end
4. Stripped trailing comments; this starts at the first unescaped `%`

Tests added for this but please let me know if something is still off.

Reply:
> spaces at the start of a line are handled by `trimws()`

Bah, thanks, I had that straight at one point, and then got it wrong with a last minute edit to my comment.